crates/amaru-ledger/src/rules.rs

```diff
@@ -100,11 +123,12 @@ pub fn validate_block(
     // using `zip` here instead of enumerate as it is safer to cast from u32 to usize than usize to u32
     // Realistically, we're never gonna hit the u32 limit with the number of transactions in a block (a boy can dream)
     for (i, transaction) in (0u32..).zip(transactions.iter()) {
-        // TODO handle `as` correctly
-        let witness_set = match witness_sets.get(i as usize) {
-            Some(witness_set) => witness_set,
-            None => continue,
-        };
+        let witness_set = witness_sets
+            .get(i as usize)
+            .ok_or(BlockValidationError::Unnamed(format!(
+                "Missing witness set for transaction index {}",
+                i
+            )))?;
```

Comment:
What's the rationale for making this an "unnamed" error instead of a validation error? It looks like it could happen if someone submitted a block which had different numbers of bodies and witness sets.

The cddl says "the length of transaction_bodies and transaction_witness_sets must be the same"; if you validate that the lengths match before this for loop, you could avoid this error by including witness_sets in the iterator:

```rust
    for (i, transaction, witness_set) in izip!(0u32.., transactions, witness_sets) {
```

Reply:
This is one of those things that's a bit unclear how to properly handle. I would generally say that anything defined in the CDDL should be enforced at deserialization, but the Pallas type itself doesn't have this enforced right now. So, in theory, it's not an error that should ever be seen at this stage in the process of validation, but at the moment it is.